demos/tutorials/Gibbs.py

```diff
@@ -0,0 +1,222 @@
+    Gibbs sampling is a Markov chain Monte Carlo (MCMC) method for
+    sampling a joint probability distribution.
+
+    Opposed to jointly sampling the distribution simultaneously, Gibbs
+    sampling samples the variables of the distribution sequentially,
+    one variable at a time. The sampling of each variable is done by
+    sampling from the conditional distribution of that variable given 
+    the values of the other variables.
+
+    This is often a very efficient way of sampling from a joint
+    distribution if the conditional distributions are easy to sample
+    from. This is one way to exploit the structure of the joint
+    distribution.
+
+"""
+# %%
+# Setup
+# -----
+# We start by importing the necessary modules
+
+import numpy as np
+from cuqi.testproblem import Deconvolution1D
+from cuqi.distribution import GaussianCov, Gamma, JointDistribution, GMRF, Laplace_diff
+from cuqi.sampler import Gibbs, Linear_RTO, Conjugate, UnadjustedLaplaceApproximation, ConjugateApprox
+
+np.random.seed(0)
+
+# %%
+# Forward model and data
+# ------------------------
+# We define the forward model and data.
+# Here we use a 1D deconvolution problem, so the forward model is linear,
+# that is:
+#
+# .. math::
+#    \mathbf{y} = \mathbf{A} \mathbf{x}
+#
+# where :math:`\mathbf{A}` is the convolution matrix, and :math:`\mathbf{x}` is the input signal.
+#
+# We load this example from the testproblem library of CUQIpy.
+
+# Model and data
+A, y_obs, probinfo = Deconvolution1D.get_components(phantom='square')
+
+# Get dimension of signal
+n = A.domain_dim
+
+# %%
+# Hierarchical Bayesian model
+# ---------------------------
+#
+# We define the following hierarchical model:
+#
+# .. math::
+#
+#     \begin{align}
+#         d &\sim \mathrm{Gamma}(1, 10^{-4}) \\
+#         l &\sim \mathrm{Gamma}(1, 10^{-4}) \\
+#         \mathbf{x} &\sim \mathrm{GMRF}(\mathbf{0}, d) \\
+#         \mathbf{y} &\sim \mathcal{N}(\mathbf{A} \mathbf{x}, l^{-1} \mathbf{I}_m)
+#     \end{align}
+#
+# where :math:`\mathbf{y}` is the observed data, and :math:`\mathbf{x}`
+# is the unknown signal. The hyperparameters :math:`d` and :math:`l` are
+# the precision of the prior distribution of :math:`\mathbf{x}` and
+# the noise, respectively.
+#
+# The prior distribution of :math:`\mathbf{x}` is a Gaussian Markov random
+# field (GMRF) with zero mean and precision :math:`d`. It can
+# be viewed as a Gaussian prior on the differences between neighboring
+# elements of :math:`\mathbf{x}`.
+#
+# In CUQIpy the model can be defined as follows:
+
+# Define distributions
+d = Gamma(1, 1e-4)
+l = Gamma(1, 1e-4)
+x = GMRF(np.zeros(n), lambda d: d[0])
+y = GaussianCov(A, lambda l: 1/l)
+
+# Combine into a joint distribution
+joint = JointDistribution([d, l, x, y])
+
+# Define posterior by conditioning on the data
+posterior = joint(y=y_obs)
+
+print(posterior)
+
+# %%
+# Gibbs Sampler
+# -------------
+#
+# The hierarchical model above has some important properties that we
```

Comment:
I think this is probably too advanced for a first demonstration of the Gibbs sampler.  Before exploiting structure and conjugacy can we have completely basic case?